src/rapids_pre_commit_hooks/lint.py

```diff
@@ -0,0 +1,236 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import bisect
+import contextlib
+import functools
+from functools import reduce
+
+
+class OverlappingReplacementsError(RuntimeError):
+    pass
+
+
+class Replacement:
+    def __init__(self, pos, newtext):
+        self.pos = pos
+        self.newtext = newtext
+
+    def __eq__(self, other):
+        if not isinstance(other, Replacement):
+            return False
+        return self.pos == other.pos and self.newtext == other.newtext
+
+    def __repr__(self):
+        return f"Replacement(pos={self.pos}, newtext={repr(self.newtext)})"
+
+
+class LintWarning:
+    def __init__(self, pos, msg):
+        self.pos = pos
+        self.msg = msg
+        self.replacements = []
+
+    def add_replacement(self, pos, newtext):
+        self.replacements.append(Replacement(pos, newtext))
+
+    def __eq__(self, other):
+        if not isinstance(other, LintWarning):
+            return False
+        return (
+            self.pos == other.pos
+            and self.msg == other.msg
+            and self.replacements == other.replacements
+        )
+
+    def __repr__(self):
+        return (
+            "LintWarning("
+            + f"pos={self.pos}, "
+            + f"msg={self.msg}, "
+            + f"replacements={self.replacements})"
+        )
+
+
+class Linter:
+    def __init__(self, filename, content):
+        self.filename = filename
+        self.content = content
+        self.warnings = []
+        self._calculate_lines()
+
+    def add_warning(self, pos, msg):
+        w = LintWarning(pos, msg)
+        self.warnings.append(w)
+        return w
+
+    def fix(self):
+        sorted_replacements = [
+            replacement
+            for warning in self.warnings
+            for replacement in warning.replacements
+        ]
+        sorted_replacements.sort(key=lambda replacement: replacement.pos)
+        if len(sorted_replacements) == 0:
+            return self.content
+
+        def check_overlap(r1, r2):
+            if r1 and r1.pos[1] > r2.pos[0]:
+                raise OverlappingReplacementsError(f"{r1} overlaps with {r2}")
+            return r2
+
+        reduce(check_overlap, sorted_replacements)
+
+        cursor = 0
+        retval = ""
```

Comment:
It might be nice to have a name for this:
```suggestion
        replaced_content = ""
```